### ISSUE TITLE
Dynamic Kubelet Settings

### ISSUE THREAD

**Issue by mtaufen** [kubernetes/kubernetes#27980]
**Updated:** Please see the proposal doc https://github.com/kubernetes/kubernetes/pull/29459 for tracking the design of this feature. I have changed the first comment in this issue to reflect progress on various milestones. 
# Milestones
## Initial
- [x] Rebase, clean up, and shepard #19837 through the code-review and merge process. This gets us a versioned configmap object for specifying the Kubelet config.
- [x] Determine naming conventions for the Kubelet configmaps.
- [x] Determine a namespace for the Kubelet configmaps. [They will go in `kube-system`]
- [x] Add an endpoint for the Kubelet to dump its config, so we have an easy way of seeing how it is configured. /configz has been updated in the example implementation to report the external type, and will suffice for now. 
- [x] Kubelet polls API server to detect config changes and reboots when it detects new config.
- [x] Add ability to set per-node config through API server (`kubelet-<node-name>`).
- [x] Finish refactor of KubeletConfiguration's path to Kubelet #29216.

## Tasks from the proposal
- [ ] Work on design until proposal is approved for implementation

### Work thus far
- Configmap: https://github.com/kubernetes/kubernetes/issues/19837
- Main Issue: https://github.com/kubernetes/kubernetes/issues/27980
- Refactor to clean up the path KubeletConfiguration takes on its way to the Kubelet: https://github.com/kubernetes/kubernetes/pull/29216
- Example implementation: https://github.com/kubernetes/kubernetes/pull/30090

**Comment by yujuhong** [comment]
> Specify the Kubelet config through the API server with a well-defined configmap object.

Will kubelet still support all the flags it has today? If not, will we have a local configmap manifest file to support standalone kubelet (i.e., no apiserver)? 

Another question: Can kubelet run with the minimal config, or it has to wait until it gets the full configmap? Currently, kubelet can still go ahead and start static pods even if apiserver is not responding. 


> Should an additional config completely replace the default, or should it consist of only fields which differ from the default, and layer these differences on top of the default config? Implementation might be easier with the former solution because, at present, there is not always a clear distinction between "not-specified, use default" and "set to the zero value" for fields in external types. The latter solution, however, might be better from a UX perspective.

IIUC and default==minimal config, the former (a complete config) has my vote. Layering is too hard to debug and negates some benefits of having a configmap object.
The UX perspective shouldn't be a problem if we let the apiserver populates the default values, like what it does for the api objects today.

**Comment by dchen1107** [comment]
> What should the "minimal config" for a Kubelet that just needs to contact the API server for its complete config look like?

Are we planning to have a "global" / "minimal" config for all Kubelets in a cluster? Where to store that minimal config? How Kubelets retrieve that? Then separate configmap object for individual node if there are change? If there are conflict values for the same flag between global and personalized one, which one should win? 

>  have it just use the default config

Where to get that default config? the default value for each flag built in? Or configured flagset passing to the binary? Or both? Or newly introduced configmap file on the node? 

> have to wait for the Kubelet to register before they know the node name for that Kubelet and that the Kubelet must be tolerant of restarts with potentially conflicting config.

Are we planning to define each config / flag with action level: restart binary? restart node? or in-place? Or we simplify it into one: restarting kubelet for now, then iterate them later on demand? I like the last one.

**Comment by mikedanese** [comment]
This only works if kubelet reads the configmap directly from the API, not from disk. @bgrant0607 had concerns with this.

**Comment by alex-mohr** [comment]
I vaguely recall similar ideas for (a) cluster deployment config, (b) core master components like apiserver/scheduler/controller-manager, and (c) addons.  I don't think we should block better kubelet config on all of the above also getting -- I'm just wondering if you have pointers to the (current) status on above or know whether there's a general plan and this is one sub-part of it.  @mikedanese might know too?

**Comment by mikedanese** [comment]
Yes the relevant issues are #19831 #12245 and #1627. This work is aligned with the second two issues I believe.

**Comment by therc** [comment]
I think this requires reopening and solving #20200 first. And I still believe that the canaries need to be picked in advance, pushed to and monitored. Internally, look at googledata pushes for an example — I'm assuming they're still around.

**Comment by vishh** [comment]
A few thoughts:

1. To bootstrap the master node we need config stored on local disk. I'd argue that disk based configs are necessary for worker nodes too because of the fact that the API server can become temporarily accessible and we want the kubelet to ideally continue managing its jobs until it can connect again with the API server.
2. Once a kubelet can successfully connect to the API server, it can then switch over to a new configs based on Config Maps.
3. For the purposes of bootstrapping we will need a `default` configuration that all nodes can use. The node name is not deterministic in all deployments and so we need a `default` config.
4. To configure each node differently, we need to support per-node configs. The obvious choice as of now is to create a ConfigMap with the same name as that of the node.
5. Having a per-node config also helps to canary config updates.
6. To keep things simple, let's not worry about merging configs. What ever config that kubelet ends up using, it will use it as-is.
6. Considering the lifecycle, the order of precedence for configs is as follows:
    * per-node ConfigMap
    * default ConfigMap
    * On-disk ConfigMap

We will still need a few flags for example to identify the location of on-disk config, kubeconfig, etc. Config Maps and flags should eventually be mutually exclusive.

**Comment by bgrant0607** [comment]
cc @kubernetes/sig-cluster-lifecycle

**Comment by bgrant0607** [comment]
ref #489

**Comment by mtaufen** [comment]
@yujuhong
For now, the Kubelet will continue to support the flags it has today. Eventually, flags will be deprecated in favor of on-disk config. I'm not 100% clear on how one sets up static pods, is that presently done using a flag? Eventually, we will have the Kubelet defer to on-disk config if it can't contact the API server, so I can imagine eventually dealing with static pods there.

I think I overloaded the word "default" and used it a little confusingly. Let's go with some new definitions. Let's have: 
- `kubelet-default`: The built-in default config that every Kubelet starts up with.
- `api-default`: The defaults provided for omitted/nil fields by the API server.
- `minimal-config`: Conceptually, what the Kubelet needs to start up and contact the API server. This probably consists of the location of the API server and credentials to access the API server, and might be provided via flags or via on-disk config. 
- `global-kubelet-configuration`: the `KubeletConfiguration` configmap set via the API server which is shared by all Kubelets in your cluster except those which have a `<node-name>-kubelet-configuration`.
- `<node-name>-kubelet-configuration`: the `KubeletConfiguration` configmap specific to the node named `<node-name>`, also set via the API server.

My question was whether to layer the `<node-name>-kubelet-configuration` on top of the `global-kubelet-configuration` or to completely replace the `global-kubelet-configuration` with the `<node-name>-kubelet-configuration`. I 100% agree on avoiding layering for the reasons you brought up, so I think we've solved this one :).

@dchen1107
I made some clarifications below, let me know if that satisifes your questions.


# More stuff
I've had a few meetings with @vishh where we've clarified how things will work and set some initial milestones. 

# Clarifications

## "Types" of config that we'll deal with, for review
- `kubelet-default`: The built-in default config that every Kubelet starts up with.
- `api-default`: The defaults provided for omitted/nil fields by the API server.
- `minimal-config`: Conceptually, what the Kubelet needs to start up and contact the API server. This probably consists of the location of the API server and credentials to access the API server, and might be provided via flags or via on-disk config. 
- `global-kubelet-configuration`: the `KubeletConfiguration` configmap set via the API server which is shared by all Kubelets in your cluster except those which have a `<node-name>-kubelet-configuration`.
- `<node-name>-kubelet-configuration`: the `KubeletConfiguration` configmap specific to the node named `<node-name>`, also set via the API server.

## Additional Goals
- It would be useful to have a Kubelet endpoint that can be used to dump its current config, so that you can check if your config was updated successfully. 
- We want the Kubelet to work with the built-in defaults on all distros we test against. 
- We want to replace flags with on-disk config. Deprecation of Kubelet flags will be an exit-criterion for the Dynamic Kubelet Config feature, but is not the primary value proposition of the feature. Until flags are deprecated, modifications to `KubeletConfiguration` need to be added as flags and to the internal and external types. 
- We will focus first on what we need in order to be able to specify different Kubelet config for tests, and then expand to more use cases.

## Idea of what bringing up a master node will look like
- Start the Kubelet with `minimal-config`
- On-disk static pod manifests are used to bring up the API server, scheduler, controller-manager, etc.
Once the master is ready, the user creates some configs via the API:
- They might create a `global-kubelet-configuration`
- They might create a `<node-name>-kubelet-configuration`, e.g. if they want a master-specific configuration.
- Then the Kubelets' configuration sync loops will see the new config, and the Kubelets will restart to take up that new config. 

## Idea of what the startup of a Kubelet will look like
- Start the Kubelet with `minimal-config`
- If `<current-node-name>-kubelet-config` exists on the API server, download and use the `<current-node-name>-kubelet-config`
- else if `global-kubelet-config` exists on the API server, download and use the `global-kubelet-config`.
- else if e.g. an `on-disk-config` flag is set, defer to the node's on-disk config. (Note: Eventually, flags will be deprecated and on-disk config will be the final fallback.)
- else defer to config from flags
- Finally, start the configuration sync loop in a background thread


# Milestones for the next two weeks
@vishh and I also came up with a few milestones to target over the next couple of weeks:

## In parallel with everything
- Group Kubelet configmap fields into substructures. @timstclair and I came up with some ideas for this last week, I will open a separate issue for this with further explaination. 

## Today
- (hopefully) #19837 goes in.

## End of this week
- Endpoint for the Kubelet to dump its config, so we have an easy way of seeing how it is configured.
- Basic functionality of starting the Kubelet with some mimimal config and downloading the `global-kubelet-configuration` from the API server to populate config.

## End of next week
- Add a flag that lets Kubelet load config from disk.
- Add ability to set per-node config through API server (`<node-name>-kubelet-configuration`).
- Kubelet polls API server to detect config changes and reboots when it detects new config.

**Comment by aaronlevy** [comment]
> Then the Kubelets' configuration sync loops will see the new config, and the Kubelets will restart to take up that new config

How does this work? Does this proposal include locally checkpointing api-defined-configs (e.g. `global-kubelet-configuration` or `<node-name>-kubelet-configuration`)?

**Comment by mtaufen** [comment]
The first iteration of this feature it's just going to be a deep comparison between what is found on the API server and what is local on the given Kubelet. We have talked about local checkpointing of config being a nice feature to have eventually, but we haven't extensively discussed how it will work. I think it might be worth opening a separate issue to discuss a general solution for checkpointing config (unless one already exists, but I haven't seen one).

**Comment by aaronlevy** [comment]
Hmm. Still a little unclear on the statement "kubelets will restart to take up that new config" -- can you describe how you envision that working?

**Comment by mtaufen** [comment]
Part of the Kubelet startup process will be checking the API server and applying the relevant config that it finds there. So when the Kubelet detects new config it can cleanly download and apply the new config by just restarting.

**Comment by aaronlevy** [comment]
@mtaufen sorry, I might be missing something.

So "cleanly download and apply the new config just by restarting". How is the config which is downloaded applied to the kubelet after restart? By restart do you mean a kubelet process restart? Or the internal kubelet object is restarted within the same process (and with a new config)?

The part I'm missing is how is this new config actually applied to the kubelet.

**Comment by mtaufen** [comment]
Ah, sure. The devil is in the details :). 

@vishh and @mikedanese and I briefly chatted in our first meeting about this feature, which was a few weeks ago, about how we might do the restart (both options you mentioned came up as possibilities). I'm hoping to have the details of at what stage, exactly, we apply the config figured out and implemented by the end of this week and the details of how we do the restart figured out and implemented by the end of next week.

If you have ideas on good ways to do these things I'd love to hear them.

**Comment by aaronlevy** [comment]
Without being privy to prior discussions, I might argue for saving the config state outside the kubelet (and reboot simply picks up this local / on-disk config).

A reason for this would be to minimize the delta of `minimal-config` to the api-defined config through lifecycle of the kubelet. A specific case that has been brought up in the past are cgroup related settings (e.g. kubelet + minimal-config starts static pods, then gets a new config with different cgroup settings, then starts more pods). Or maybe that's just covered by a "don't do that" clause.

**Comment by mtaufen** [comment]
That sounds very reasonable to me. This is probably a good reason to eventually do local checkpointing of config. I'd like to keep things as simple as possible for the next two weeks so I can get a basic prototype of the feature working sooner rather than later. 

I think this is important to discuss further. Is there a current issue related to checkpointing config? I see #489, but that looks more geared towards checkpointing pod manifests. If that's not the appropriate place for talking about config checkpointing, I'll open a separate issue and we can start figuring out how we want to do it there.

**Comment by mikedanese** [comment]
We need to verify that the round trip conversion tests are running on this api group, and enable them if they are not.

**Comment by derekwaynecarr** [comment]
Is it possible we can get a high-level proposal doc written so we can comment on particular areas?  Right now, the issue is kind of a blended set of proposals, and it would be good if we can have a PR that iterates to capture agreement on things.

**Comment by mtaufen** [comment]
@derekwaynecarr I think that's a good idea. It will be a good opportunity to go through and organize everything that has been said so far as well. I will start working on something this afternoon.

**Comment by mtaufen** [comment]
First draft of proposal here: https://github.com/kubernetes/kubernetes/pull/29459

**Comment by mtaufen** [comment]
Tracking an example implementation of some basic functionality here: #30090.

**Comment by resouer** [comment]
cc @resouer Just tracking status of this.

**Comment by ncdc** [comment]
FYI we just created #32215 with some suggestions for how to improve the configuration structs and process a bit. If we get agreement to move forward with those, we'll want the Kubelet config to follow suit. I'd really appreciate it if you could take a look at the suggestions and weigh in with your thoughts. Thanks!

**Comment by dims** [comment]
switching this to 1.6 as it's too late for 1.5. ok? (please switch it right back if you disagree)

**Comment by luxas** [comment]
We really need an assignee for this for v1.6
Will someone from @kubernetes/sig-node work on this in time for v1.6?

The base config that applies to all kubelets but can be overridden will be a critical component for us in @kubernetes/sig-cluster-lifecycle

**Comment by mtaufen** [comment]
I am planning on picking this back up for 1.6.

**Comment by luxas** [comment]
Thanks, the sooner we get it the better.
Feel free to ping me on any of those issues/PRs.

Do you think we should support layering or should we just provide the basic base configmap for all kubelets and make all `kubelet-nodename` configmaps override the base configmap completely?

I think that might be one of the largest design issues we have to make
cc @vishh

**Comment by vishh** [comment]
> Do you think we should support layering or should we just provide the
basic base configmap for all kubelets and make all kubelet-nodename configmaps
override the base configmap completely?

Kubelet-nodename will not work. We need to take into account upgrades and
rollbacks. @mtaufen can you post the design updates from our offline
discussions?

**Comment by mtaufen** [comment]
@vishh and I had those discussions several months ago, I need to dig back into my notes for exactly what we talked about.

It's probably time to think about the general case of component (configuration) life-cycle. I would rather determine good semantics for component configuration, and then implement these semantics in the Kubelet as a prototype, vs. making something that ends up kubelet-specific and hard to implement for other components.

@luxas people keep asking for layering, so it probably makes sense at some point. The original reason that we avoided any sort of aggregate config and decided to force per-node configuration was that it was too easy to hose a cluster by pushing a node-specific parameter in global config; this *could* be avoided by splitting componentconfig types into parameters that only make sense on a per-object basis and parameters that can be safely applied to groups of objects, but we haven't had that discussion yet. The best structure of these types is yet-to-be-determined and is in the queue of design discussions that need to be had for 1.6.

**Comment by luxas** [comment]
At least @kubernetes/sig-cluster-lifecycle needs the base kubelet layer for a lot of things before we can move on.

It's a blocker for us that we have to get rid of soon, in the v1.6 timeframe preferably.
See: https://github.com/kubernetes/kubeadm/issues/28

The main pain points is that we have no way to let the configuration passed to the apiserver flow down to the kubelets right now. For example with cloudprovider and dns settings, we can just say to users now: "Go configure all your kubelets manually", but that shouldn't be the case if we had this functionality

I really look forward to getting this solved by discussing the design and then implementing it

**Comment by mtaufen** [comment]
@vishh as I am able to recall after digging through my notes, our discussion a few months ago focused on rollouts/rollbacks along these lines:
- Have a controller roll out config by annotating nodes
- Track Kubelet restarts over a period of time, if these exceed a (TBD) threshold, revert to the previous config
- Perhaps add something like a "LastRestartReason" to the Kubelet and track this as well - e.g. to make it easy for the Kubelet to report "invalid config"
- And we also that we really need to separate non-generalizable parameters from generalizable parameters

I am currently reading through the old GitHub discussions and the new, related PRs and Issues that have sprung up in the last quarter or so. Once I have more up-to-date context and get my thoughts decently together on this (today or Monday), I will post a more detailed comment wrt what I think we need to consider.

**Comment by mtaufen** [comment]
Just posted a comment with my thoughts over in the proposal PR: https://github.com/kubernetes/kubernetes/pull/29459#issuecomment-264961734

Let's keep the conversation focused over there.

**Comment by bogdando** [comment]
Cross linked as it's related to kubelet config changes https://github.com/kubernetes/kubernetes/issues/12245#issuecomment-270075340

**Comment by cmluciano** [comment]
@mtaufen I can help out with any additional work related to this epic. Are there specific pieces that you still need help with?

**Comment by mtaufen** [comment]
@cmluciano very sorry I missed your comment. If you're still interested in helping out the majority of the discussion is taking place in #29459. I expect the implementation of the foundational bits will be relatively lightweight work, but it would be really useful to have another hand when we get to building orchestration tooling. I understand if it's too late for you at this point :).

**Comment by cmluciano** [comment]
I totally understand  @mtaufen . Taking a look at that PR

**Comment by mtaufen** [comment]
Alpha PR: https://github.com/kubernetes/kubernetes/pull/46254

**Comment by mtaufen** [comment]
Closing this issue, as the feature is now tracked via https://github.com/kubernetes/features/issues/281
